### PR TITLE
 Adjust ListT to reflect successive base implementations of Monoid / Semigroup

### PR DESCRIPTION
**Pull request by jrp2014**
Refactoring.  Should make this code easier to clean up when GHC 8.4 is no longer supported.

### CODE REVIEW

**Comment by andreasabel** [comment]
> Refactoring. Should make this code easier to clean up when GHC 8.4 is no longer supported.

Ok, but this is still far off in the future...

**Review by andreasabel — APPROVED** [review]
LGTM, though I do not understand all subtleties, do you have a reference for the INLINE, is there a specific policy?

**Comment by jrp2014** [comment]
[Inilining and Specialisation](https://mpickering.github.io/posts/2017-03-20-inlining-and-specialisation.html) is not very detailed but I note that Ed Kmett, Ryan Scott et al seem to use them liberally in this context.   There may be other basic building blocks that might benefit.

Incidentally, one reason for trying to generate explicit export lists is that the manual says:

> If you do not have an explicit export list in a module, GHC must assume that everything in that module  will be exported. This has various pessimising effects. For example, if a bit of code is actually unused (perhaps because of unfolding effects), GHC will not be able to throw it away, because it is exported and some other module may be relying on its existence.


I'd be quite interested to see how old the GHCs that most people use are.  Presumably stack users don't care.

**Comment by andreasabel** [comment]
#1972 is an example discussion about removing support for GHC versions (there: 7.6).

**Comment by andreasabel** [comment]
> Inilining and Specialisation 

This post contains the following explanation:

> There are two ways to make functions specialise across modules:
>
> 1. Marking a function as INLINABLE or INLINE.
> 2. Using the flag -fspecialise-aggressively when compiling the client module. An unfolding must still be available to perform specialisation.

We are using the flag `-fspecialise-aggressively`.  What do `INLINE` annotations buy us in addition?

On export lists:
> For example, if a bit of code is actually unused (perhaps because of unfolding effects), GHC will not be able to throw it away, because it is exported and some other module may be relying on its existence.

Is there great harm in this?  This will make the executable bigger, but that isn't a major bottleneck.

On the other hand, export lists lead in development practice to productivity loss: Forgetting to update the export list with a new function might be discovered only 200 modules down the pipeline (at import time).  Even with `make quicker-install-bin` I am loosing several minutes on each failed compilation attempt, and the feeling of flow.

**Comment by jrp2014** [comment]
I'll zap the inlining.

This is probably out of date https://wiki.haskell.org/Performance/Modules but some of the other documentation suggests that unexported functions can be better optimized.

I agree that compile times are a bottleneck.  I tend to go with `cabal repl`  but even that could probably be speeded up by switching off code generation.  Its key benefit when working on a small set of files is that reloading is fast.  A different approach would be to use ghcid although I didn't find it very robust when applied to the Agda codebase.

File: src/full/Agda/Utils/ListT.hs [hunk 1]
```diff
@@ -109,9 +109,13 @@ liftListT lift xs = runMListT $ caseMaybeM (lift $ runListT xs) (return nilListT
 instance Monad m => Semigroup (ListT m a) where
   l1 <> l2 = ListT $ foldListT cons (runListT l2) l1
     where cons a = runListT . consListT a . ListT
+
 instance Monad m => Monoid (ListT m a) where
+  {-# INLINE mempty #-}
```
Comment (andreasabel) [inline src/full/Agda/Utils/ListT.hs]: Did you mean to zap this?

Comment (jrp2014) [reply src/full/Agda/Utils/ListT.hs]: Argh.  Yes.  (I need to clean up my vim plug-ins; the summarising plug-in is not always your friend.)

**Comment by andreasabel** [comment]
Thanks, @jrp2014.

**Pull request merged**